tests/unittests/test_datasource/test_ec2.py

```diff
@@ -331,6 +433,37 @@ def test_network_config_property_set_dhcp4_on_private_ipv4(self):
                     m_get_mac.return_value = mac1
                     self.assertEqual(expected, ds.network_config)
 
+    def test_network_config_property_secondary_private_ips(self):
+        """network_config property configures any secondary ipv4 addresses.
+
+        Only one device is configured even when multiple exist in metadata.
+        """
+        ds = self._setup_ds(
+            platform_data=self.valid_platform_data,
+            sys_cfg={'datasource': {'Ec2': {'strict_id': True}}},
+            md={'md': SECONDARY_IP_METADATA_2018_09_24})
+        find_fallback_path = M_PATH_NET + 'find_fallback_nic'
+        with mock.patch(find_fallback_path) as m_find_fallback:
+            m_find_fallback.return_value = 'eth9'
+            ds.get_data()
+
+        mac1 = '0a:07:84:3d:6e:38'  # IPv4 with 1 secondary IP
+        expected = {'version': 2, 'ethernets': {'eth9': {
+            'match': {'macaddress': mac1}, 'set-name': 'eth9',
+            'addresses': ['172.31.45.70/20',
+                          '2600:1f16:292:100:f153:12a3:c37c:11f9/128',
+                          '2600:1f16:292:100:f152:2222:3333:4444/128'],
+            'dhcp4': True, 'dhcp6': True}}}
+        patch_path = M_PATH_NET + 'get_interfaces_by_mac'
+        get_interface_mac_path = M_PATH_NET + 'get_interface_mac'
+        with mock.patch(patch_path) as m_get_interfaces_by_mac:
+            with mock.patch(find_fallback_path) as m_find_fallback:
+                with mock.patch(get_interface_mac_path) as m_get_mac:
+                    m_get_interfaces_by_mac.return_value = {mac1: 'eth9'}
+                    m_find_fallback.return_value = 'eth9'
+                    m_get_mac.return_value = mac1
+                    self.assertItemsEqual(expected, ds.network_config)
```

Comment:
I have some reservations about using `assertItemsEqual` here. The error message is showing what the difference is:

```diff
+ {'ethernets': {'eth9': {'addresses': ['2600:1f16:292:100:f153:12a3:c37c:11f9/128',
- {'ethernets': {'eth9': {'addresses': ['172.31.45.70/20',                      
-                                       '2600:1f16:292:100:f153:12a3:c37c:11f9/128',
-                                       '2600:1f16:292:100:f152:2222:3333:4444/128'],
?                                                                                  -                  
                                                                                
+                                       '2600:1f16:292:100:f152:2222:3333:4444/128',
+                                       '172.31.45.70/20'],                     
                          'dhcp4': True,                                        
                          'dhcp6': True,                                        
                          'match': {'macaddress': '0a:07:84:3d:6e:38'},         
                          'set-name': 'eth9'}},                                 
```

Specifically, `addresses` is not in the expected order. Is it the case that the order of `addresses` truly doesn't matter?

There's at least one case in the code where the ordering will change behaviour (though I don't know if that change is meaningful): https://github.com/canonical/cloud-init/blob/65a1b907c336786bce3917fad3f87c67f0caa7bf/cloudinit/net/network_state.py#L807-L810

Reply:
I've added a comment the the affect that this ordering doesn't matter because rendering puts this config up in the interface scope instead of tied to the individual subnet.
https://github.com/canonical/cloud-init/blob/master/tests/unittests/test_net.py#L2061-L2099

Reply:
I think this comment may be the only one in contention. Ryan messaged in IRC saying he wasn't concerned about the ordering because we add all routes to the first subnet regardless of which subnet that ends up being in v1. And since v2 is all at the interface, it really is a don't care at all.

Reply:
https://github.com/canonical/cloud-init/pull/114#discussion_r392419963